Restructure documentation (#117)
IAMconsortium/nomenclature: doc/source/user_guide/codelist.rst

.. _codelist:

> Codelist

Reviewer: ```suggestion
Validation using Codelists
```

========

> A "codelist" (named in reference to code lists in `SDMX <https://sdmx.org/>`_) is a yaml

Reviewer: A codelist is a list of codes. In this package, the codelists are parsed from yaml files.

file that contains a list of allowed values for any **index** dimension of the IAMC
format (model, scenario, region, variable). Which index dimension a code list will
be applied to is determined by the **name of the folder** in which it is located. For
example, if a codelist is located in a folder called ``model/`` it will be applied to
the "model" index dimension. 

> A codelist must only contain a single list of allowed values. In the most simple case

Reviewer: What is the message here?
"A codelist must only contain a single list of allowed values."

Author: The message was supposed to be that a codelist can only be a single list meaning something like this:

```yaml
allowed_regions:
  - region_a
  - region_b
  - ...
allowed_variables:
  - variable_a
  - ...
```

would not be allowed as this would be two lists (which are part of a mapping).
But you're right, the sentence probably does more harm than good and I think it can be removed without loss of information.

the list items are strings, e.g.:

.. code:: yaml

  - allowed_value_a
  - allowed_value_b
  - allowed_value_c
  - ...

In case of the "model" or "scenario" index dimension, such a simple list sufficient.

> More details on 

Reviewer: This sentence is not complete. However, I would omit this entire paragraph. Users will scroll through the page for the heading that they are interested in (variables, regions, ...).

Author: Sounds good, removed it.

For defining allowed variables and regions, however, the list elements are more complicated, detailed in the following.

.. _variable:

Variable
--------

An entry in a variable code list, **must be** a mapping (translated to python as a

> dictionary) maps the **name** of an allowed variable to, at minimum, a key value pair

Reviewer: ..must be a mapping maps the name..

Author: Corrected the sentence.

   *avg*) but always spell out the word.
*  All words must be capitalised (except for *and*, *w/*, *w/o*, etc.).
*  Add hierarchy levels where it might be useful in the future, e.g.,
   use *Electric Vehicle|Plugin-Hybrid* instead of *Plugin-Hybrid
   Electric Vehicle*.
*  Do not include words like *Level* or *Quantity* in the variable,
   because this should be clear from the context or unit.

Region
------

As is the case for the "variable" codelist, a region codelist must also follow a
specific structure. 

Each region **must** be part of a hierarchy, which means that the following nested list
structure is required:

.. code:: yaml

   - Hierarchy 1:
     - region 1:
        some attribute: some value
     - region 2
   - Hierachy 2:
     - ...  

Attributes of the **regions**, in the above example *some attribute* of *region 1* are
entirely optional and have no influence on the region-processing.

Nonetheless, they can be very useful, examples are: ISO2/3-codes
(https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes) or the list of countries
included in a macro-region (i.e., a continent or large region).

.. _generic:

Generic
-------

> As mentioned in the beginning of this section, for IAMC dimensions other than 'region'

Reviewer: ```suggestion
For IAMC dimensions other than 'region'
```

.. _codelist:

Validation using Codelists
==========================

A codelist is a list of codes (i.e. allowed values). In this package, the codelists are
parsed from yaml files. They contain lists of allowed values for any **index**
dimension of the IAMC format (model, scenario, region, variable). Which index dimension
a codelist will be applied to is determined by the **name of the folder** in which it is
located. For example, if a codelist is located in a folder called ``model/`` it will be
applied to the "model" index dimension. 

Codelist format specification
-----------------------------

> In the most simple case of a codelist it contains a list of strings, e.g.:

Reviewer: ```suggestion
In the most simple case, a codelist consists of a list of strings, e.g.:
```

.. code:: yaml

  - allowed_value_a
  - allowed_value_b
  - allowed_value_c
  - ...

For defining allowed *variables* and *regions*, the list elements are more complicated
than simple strings, detailed in the following.

.. _variable:

Variable
^^^^^^^^

An entry in a variable code list, *must be* a mapping (translated to python as a
dictionary). It maps the **name** of an allowed variable to, at least, one key-value
pair defining the allowed **unit(s)** for the variable.

This is an example for a valid entry in a variable codelist:

.. code:: yaml

   - Allowed variable name:
     unit: A unit
     description: A short description
     <other attribute>: Some text, value, boolean or list (optional)

The **unit** attribute is **required** and its value should be compatible with the
`iam-units <https://github.com/iamconsortium/units>`_ package.

> The unit attribute can be:

Reviewer: This is not formatted as a list on RTD...

Author: Added a second line break, now it's rendered correctly.

* a string -> one allowed unit for the variable
* a list of strings -> a number of allowed units for the variable 
* empty -> a *dimensionless* variable

examples for all three options:

    .. code:: yaml
      
      - Single unit variable:
        unit: A single unit
      - Multi unit variable:
        unit: [unit 1, unit 2]
      - Dimensionless variable:
        unit:      

While not strictly necessary a *description* attribute with a short description of the
variable is encouraged. 

> In principle, *any* number of additional arbitrary named attributes are allowed. There
> are several attributes that affect the region-processing by the nomenclature package
> (see :ref:`region_aggregation_attributes`).

Reviewer: ```suggestion
The yaml format allows *any* number of additional arbitrary named attributes. 
```

.. _region_aggregation_attributes:

Optional attributes for region aggregation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> 

Reviewer: ```suggestion
There  are several attributes that affect the region-processing by the nomenclature package.
See the Section :ref:`model_mapping` for more information.
```

* By default, all variables are processed using the method
  :meth:`pyam.IamDataFrame.aggregate_region`, which performs a simple summation of all
  subregions.

* Region aggregation for a particular variable can be skipped by using the attribute
  *skip-region-aggregation* and setting it to ``true``:

    .. code:: yaml

       - Some Variable:
         skip-region-aggregation: true

>   Setting *skip-region-aggregation* to ``true`` only skips the variable in question for

Reviewer: This sentence would be easier to understand if you cut it into two sentences.

Author: Good point, split in two.

  aggregation, if the variable is part of the provided data, it **is** used.

* Any attributes which are arguments of
  :meth:`aggregate_region() <pyam.IamDataFrame.aggregate_region>` will
  be passed to that method. Examples include *method* and *weight*.

* It is possible to rename the variable returned by the region processing using
  a *region-aggregation* attribute, which must have a mapping of the target variable to
  arguments of :meth:`aggregate_region() <pyam.IamDataFrame.aggregate_region>`.

  This option can be used to compute several variables as part of the region-processing.
  In the example below, the variable *Price|Carbon* is computed as a weighted average
  using the CO2 emissions as weights, and in addition, the maximum carbon price within
  each aggregate-region is added as a new variable *Price|Carbon (Max)*.

    .. code:: yaml

        - Price|Carbon:
            unit: USD/t CO2
            region-aggregation:
              - Price|Carbon:
                  weight: Emissions|CO2
              - Price|Carbon (Max):
                  method: max

> Guidelines and variable naming conventions

Reviewer: I suggest to move this "Guideline" section above the section "Optional attributes for region aggregation", because it is "more important" (everyone should be aware of these guidelines, users don't necessarily need to know about the region-aggregation attributes...

Author: Switched the two.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The variable name should adhere to the following conventions:

*  A *|* (pipe) character indicates levels of hierarchy.
*  Do not use spaces before and after the *|* character, but add a
   space between words (e.g., *Primary Energy|Non-Biomass Renewables*).
*  Do not use abbreviations (e.g, *PHEV*) unless strictly necessary.
*  Do not use abbreviations of statistical operations (*min*, *max*,
   *avg*) but always spell out the word.
*  All words must be capitalised (except for *and*, *w/*, *w/o*, etc.).
*  Add hierarchy levels where it might be useful in the future, e.g.,
   use *Electric Vehicle|Plugin-Hybrid* instead of *Plugin-Hybrid
   Electric Vehicle*.
*  Do not include words like *Level* or *Quantity* in the variable,
   because this should be clear from the context or unit.

Region
^^^^^^

> As is the case for the "variable" codelist, a region codelist must also follow a
> specific structure. 
> 

Reviewer: This sentence does not add relevant information.
```suggestion
```

Each region **must** be part of a hierarchy, which means that the following nested list
structure is required:

.. code:: yaml

   - Hierarchy 1:
     - region 1:
        some attribute: some value
     - region 2
   - Hierachy 2:
     - ...  

> Attributes of the **regions**, in the above example *some attribute* of *region 1* are
> entirely optional and have no influence on the region-processing.
> 
> Nonetheless, they can be very useful, examples are: ISO2/3-codes
> (https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes) or the list of countries
> included in a macro-region (i.e., a continent or large region).

Reviewer: Saying that something is unrelated to something else can be confusing if users wouldn't even have thought that they are related in the first place...
```suggestion
Useful examples of region attriutes are: ISO2/3-codes
(https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes)
or the list of countries included in a macro-region (i.e., a continent).
```

.. _generic:

Generic
^^^^^^^

For IAMC dimensions other than 'region'
and 'variable' (e.g. 'scenario' or 'model'), the requirements for are more simple: 

.. code:: yaml

   - scenario 1
   - scenario 2:

>      description: Something about scenario 1

Reviewer: This needs to be indented, right?

   dsd = DataStructureDefinition("definition")
   # data to validate in IAMC format
   data = pyam.IamDataFrame("input_data.xlsx") 
   
   # returns True if the data is valid, raises error otherwise
   dsd.validate(data)

Per default, :class:`DataStructureDefinition` reads in *region* and *variable* codelists
from their respective sub folders inside the ``definition/`` folder. Any different
number of dimensions can be read in by instantiating the ``DataStructureDefinition``
object with an additional list of strings, e.g. ``DataStructureDefinition("definition",
['region', 'variable', 'scenario'])``. This would attempt to read three codelists.

In addition, when running :meth:`DataStructureDefinition.validate`, it can be selected
which dimensions to *validate*. Per default, *all* dimensions which were read at
instantiating are validated, but any subset can be selected by providing a list of
dimensions. In the above example using ``dsd.validate(df, ['scenario'])`` would validate
*only* the *scenario* dimension.

In practice, ``DataStructureDefinition.validate`` is usually not called directly but
rather as part of the :func:`process` function which combines validation and region

> processing. Further details on that are found in :ref:`minimum_working_example`.

Reviewer: ```suggestion
processing. Refer to the :ref:`minimum_working_example` for details.
```
